### PR TITLE
Feature/esp button

### PR DESCRIPTION
**Pull request by alibukharai**
PTAL

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,3 +1,171 @@
 # ESP32_Button
 
-Arduino library of driving button for the ESP32 SoCs
+ESP32_Button is an Arduino library designed for creating both GPIO and ADC buttons using ESP32 System-on-Chip (SoC) devices. This versatile library simplifies the process of configuring and working with buttons on ESP32-based projects. It builds upon the functionality provided by the [ESP-IOT-SOLUTION/Components/Button](https://github.com/espressif/esp-iot-solution/tree/master/components/button) library, offering an accessible interface for button-related tasks in the Arduino ecosystem.
+
+## Supported Events
```
Comment (Lzw655) [inline README.md]: Please refer to [the README in template](https://github.com/esp-arduino-libs/template_project/blob/master/README.md) here and it should at least contain `Features` and so on. And I think the content here and below belong to `How to Use` - `Detailed Usage`.

File: src/Button.h [hunk 2]
```diff
@@ -0,0 +1,81 @@
+/*
+ * SPDX-FileCopyrightText: 2023-2024 Espressif Systems (Shanghai) CO LTD
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef BUTTON_H
+#define BUTTON_H
+#include <Arduino.h>
```
Comment (Lzw655) [inline src/Button.h]:
```suggestion

```

File: examples/example/example.ino [hunk 3]
```diff
@@ -0,0 +1,47 @@
+/*
+ * SPDX-FileCopyrightText: 2023-2024 Espressif Systems (Shanghai) CO LTD
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
```
Comment (Lzw655) [inline examples/example/example.ino]:
```suggestion
#include "Arduino.h"
```
The Arduino project doesn't need to contain copyright.

File: library.properties [hunk 4]
```diff
@@ -0,0 +1,9 @@
+name=ESP32_Button
+version=1.0.0
```
Comment (Lzw655) [inline library.properties]:
```suggestion
version=0.0.1
```

File: library.properties [hunk 5]
```diff
@@ -0,0 +1,9 @@
...
+author=Ali Hassan shah <ali.shah@espressif.com>
```
Comment (Lzw655) [inline library.properties]:
```suggestion
author=espressif
```
This filed will be shown in Arduino IDE. So using the company's name is better.

File: library.properties [hunk 6]
```diff
@@ -0,0 +1,9 @@
...
+maintainer=Ali Hassan shah <ali.shah@espressif.com>
+sentence=This is an Arduino wrapper that enables the integration of a button component in an IoT solution.
+paragraph=This library implements the Button Component from Iot-Solutions SDK for ESP32 on the Arduino platform. It allows easy integration of buttons into IoT projects. With this library, users can quickly add button functionality to their ESP32-based IoT devices. It simplifies the development process and provides support for ESP32-based IoT projects using the Arduino platform.
```
Comment (Lzw655) [inline library.properties]: Please refer to the [repo ](https://github.com/esp-arduino-libs/ESP32_Knob/blob/master/library.properties)and [template](https://github.com/esp-arduino-libs/template_project/blob/a556b1e03ee33f2c262bc12d2307185d3b8a63b1/library.properties#L5-L6).

File: library.properties [hunk 7]
```diff
@@ -0,0 +1,9 @@
+name=ESP32_Button
+version=1.0.0
+author=Ali Hassan shah <ali.shah@espressif.com>
+maintainer=Ali Hassan shah <ali.shah@espressif.com>
+sentence=This is an Arduino wrapper that enables the integration of a button component in an IoT solution.
+paragraph=This library implements the Button Component from Iot-Solutions SDK for ESP32 on the Arduino platform. It allows easy integration of buttons into IoT projects. With this library, users can quickly add button functionality to their ESP32-based IoT devices. It simplifies the development process and provides support for ESP32-based IoT projects using the Arduino platform. 
+category=Other
+url=https://github.com/espressif/esp-iot-solution/tree/master/components
```
Comment (Lzw655) [inline library.properties]:
```suggestion
url=https://github.com/esp-arduino-libs/ESP32_Button
```

File: src/Button.h [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+
+    // Define the Button class
+class Button {
+
+public:
+    // Constructors for gpio button
+    Button(int pin, const bool pullup);
+
+    // Constructors for gpio button
+    Button(int pin, const bool pullup, uint8_t adc_channel, uint8_t button_index, uint16_t min, uint16_t max);
```
Comment (Lzw655) [inline src/Button.h]:
```suggestion
    Button(int pin, bool pullup);

    // Constructors for gpio button
    Button(int pin, bool pullup, uint8_t adc_channel, uint8_t button_index, uint16_t min, uint16_t max);
```
`const` should be used with the pointer.

File: src/Button.h [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+    void buttonDelete(void);
+    int countCallBack(void);
+    int countEvent(button_event_t event);
+    int getEvent(void);
+    u_int8_t getRepeat(void);
+    u_int16_t getTickTime(void);
+    u_int16_t getLongPressHoldCount(void);
+    void setParam(button_param_t param, void *value);
+    void buttonResume(void);
+    void buttonStop(void);
```
Comment (Lzw655) [inline src/Button.h]:
```suggestion
    void resume(void);
    void stop(void);
    void delete(void);
    int getCallbackCount(void);
    int getEventCount(button_event_t event);
    int getEvent(void);
    int getClickRepeatCount(void);
    int getTickTime(void);
    int getLongPressHoldCount(void);
    void setParam(button_param_t param, void *value);
```
Please consider them. And modify the source code if accepted.

File: src/Button.h [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+
+private:
+    // Private variables
+    int _buttonpin;
```
Comment (Lzw655) [inline src/Button.h]:
```suggestion
    int _button_pin;
```

**Review by Lzw655 — COMMENTED** [review]
Please refer to the [Konb repo](https://github.com/esp-arduino-libs/ESP32_Knob) to add `test_apps` and `.github`.

**Comment by Lzw655** [comment]
@alibukharai I tried to build the `example.ino` but got many errors. Did you try this before?

**Comment by alibukharai** [comment]
> @alibukharai I tried to build the `example.ino` but got many errors. Did you try this before?

Yes, I just checked, and it appears that some headers are missing. To resolve the issue, we should make the following changes:

1. Include `<Arduino.h>` in a code because we are using constants like "Pullup" from it. 

   ```cpp
   #include <Arduino.h>
   ```

2. In  `.cpp` file, we need to include the `"Button.h"` header at the beginning. This ensures that the class definition is visible to the implementation code:

   ```cpp
   #include "Button.h"
   ```

These changes should resolve the compilation errors. I will push the changes with test app on monday. Happy weekend :)

File: README.md [hunk 11]
```diff
@@ -1,3 +1,196 @@
...
+
+* [example.ino](./examples/example/example.ino): Demonstrates how to use ESP32-Button.
+
+
+## Detailed Usage
+
+### Initializing a GPIO Button
+```
+  Button *btn = new Button(pin number, pullupActive)
+
```
Comment (Lzw655) [inline README.md]: Remove these blank lines?

File: README.md [hunk 12]
```diff
@@ -1,3 +1,196 @@
...
+- **pin :** which specifies the pin number that the button is connected to,
+
+- **pullupActive :** which is a boolean value indicating whether the internal pull-up resistor of the pin should be enabled or not. If pullupActive is true, the pinMode() function sets the pin as an input with its internal pull-up resistor enabled. This means that the pin will read as HIGH when the button is not pressed, and LOW when the button is pressed. If pullupActive is false, the pinMode() function sets the pin as an input with its internal pull-up resistor disabled. In this case, the pin will read as a floating state, which may be either HIGH or LOW when the button is not pressed. When the button is pressed, it will again momentarily connect the pin to ground, causing it to read as LOW.
+
+- **adc_channel :** Channel of ADC
+
+- **button_index :** button index on the channel
+
+- **min :** min voltage in millivolt corresponding to the button
+
```
Comment (Lzw655) [inline README.md]: Also here.

File: test_apps/main/button_test.cpp [hunk 13]
```diff
@@ -0,0 +1,105 @@
...
+{
+    ESP_LOGI(TAG, "Button single click");
+}
+static void onButtonSingleClickRepeatCb(void *button_handle, void *usr_data)
+{
+    ESP_LOGI(TAG, "Button single click repeat");
+}
+
+
+TEST_CASE("gpio button test", "[button][iot]")
```
Comment (Lzw655) [inline test_apps/main/button_test.cpp:58]: Also add the original case in test_apps?

**Review by Lzw655 — APPROVED** [review]

**Pull request merged**